generator/restriction_writer.cpp

```diff
@@ -62,30 +90,59 @@ void RestrictionWriter::Write(RelationElement const & relationElement)
 
   CHECK_EQUAL(relationElement.GetType(), "restriction", ());
 
-  // Note. For the time being only line-point-line road restriction is supported.
-  if (relationElement.nodes.size() != 1 || relationElement.ways.size() != 2)
-    return;  // Unsupported restriction. For example line-line-line.
-
-  // Extracting osm ids of lines and points of the restriction.
-  auto const findTag = [](std::vector<std::pair<uint64_t, std::string>> const & members,
-                          std::string const & tag) {
-    auto const it = std::find_if(
-        members.cbegin(), members.cend(),
-        [&tag](std::pair<uint64_t, std::string> const & p) { return p.second == tag; });
-    return it;
+  auto const getMembersByTag = [&relationElement](std::string const & tag) {
+    std::vector<RelationElement::Member> result;
+    for (auto const & member : relationElement.ways)
+    {
+      if (member.second == tag)
+        result.emplace_back(member);
+    }
+
+    for (auto const & member : relationElement.nodes)
+    {
+      if (member.second == tag)
+        result.emplace_back(member);
+    }
+
+    return result;
   };
 
-  auto const fromIt = findTag(relationElement.ways, "from");
-  if (fromIt == relationElement.ways.cend())
-    return;
+  auto const getType = [&relationElement](uint64_t osmId) {
+    for (auto const & member : relationElement.ways)
+    {
+      if (member.first == osmId)
+        return OsmElement::EntityType::Way;
+    }
 
-  auto const toIt = findTag(relationElement.ways, "to");
-  if (toIt == relationElement.ways.cend())
-    return;
+    for (auto const & member : relationElement.nodes)
+    {
+      if (member.first == osmId)
+        return OsmElement::EntityType::Node;
+    }
+
+    UNREACHABLE();
+  };
 
-  if (findTag(relationElement.nodes, "via") == relationElement.nodes.cend())
+  auto const from = getMembersByTag("from");
+  auto const to = getMembersByTag("to");
+  auto const via = getMembersByTag("via");
+
+  if (from.size() != 1 || to.size() != 1 || via.empty())
     return;
 
+  // Either 1 node as via, either several ways as via.
```

Comment:
`// Either single node is marked as via or one or more ways are marked as via.`
Right? Several means two, three or more.

Reply:
done